shared/src/cell_set.rs

```diff
@@ -52,28 +87,36 @@ impl CellSet {
         }
     }
 
-    fn rollback(&mut self, inputs: Vec<OutPoint>, outputs: Vec<H256>) {
-        for h in outputs {
-            self.remove(&h);
-        }
+    pub fn update(&mut self, diff: CellSetDiff) {
+        diff.old_outputs.iter().for_each(|h| {
+            self.remove(h);
+        });
 
-        for o in inputs {
-            self.mark_live(&o);
-        }
-    }
+        diff.old_inputs.iter().for_each(|o| {
+            self.mark_live(o);
+        });
 
-    fn forward(&mut self, inputs: Vec<OutPoint>, outputs: Vec<(H256, usize)>) {
-        for (hash, len) in outputs {
-            self.insert(hash, len);
-        }
+        diff.new_outputs.iter().for_each(|(hash, len)| {
+            self.insert(hash, *len);
+        });
 
-        for o in inputs {
-            self.mark_dead(&o);
-        }
+        diff.new_inputs.iter().for_each(|o| {
+            self.mark_dead(o);
+        });
     }
+}
```

Comment:
`fn insert` should keep `move semantics`
 pub fn insert(&mut self, hash: H256, outputs_len: usize) 

leverage destructuring assignment in `fn update` keeps logic same as original code, avoid unnecessary clone.
```rust 
        let CellSetDiff {
            old_inputs,
            old_outputs,
            new_inputs,
            new_outputs,
        } = diff;

        old_outputs.iter().for_each(|h| {
            self.remove(h);
        });

        old_inputs.iter().for_each(|o| {
            self.mark_live(o);
        });

        new_outputs.into_iter().for_each(|(hash, len)| {
            self.insert(hash, len);
        });

        new_inputs.iter().for_each(|o| {
            self.mark_dead(o);
        });
```